docs/architecture/adr-038-state-listening.md

```diff
@@ -485,99 +307,67 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx
-
-// ServiceType enum for specifying the type of StreamingService
-type ServiceType int
-
-const (
-  Unknown ServiceType = iota
-  File
-  // add more in the future
-)
-
-// NewStreamingServiceType returns the streaming.ServiceType corresponding to the provided name
-func NewStreamingServiceType(name string) ServiceType {
-  switch strings.ToLower(name) {
-  case "file", "f":
-    return File
-  default:
-    return Unknown
-  }
+	// Closer interface for shutting down the plugin process
+	io.Closer
 }
+```
 
-// String returns the string name of a streaming.ServiceType
-func (sst ServiceType) String() string {
-  switch sst {
-  case File:
-    return "file"
-  default:
-    return ""
-  }
-}
+The `Name` method returns a plugin's name.
+The `Version` method returns a plugin's version.
+The `Init` method initializes a plugin with the provided `AppOptions`.
+The io.Closer is used to shut down the plugin service.
 
-// ServiceConstructorLookupTable is a mapping of streaming.ServiceTypes to streaming.ServiceConstructors
-var ServiceConstructorLookupTable = map[ServiceType]ServiceConstructor{
-  File: NewFileStreamingService,
-}
+For the purposes of this ADR we introduce a single kind of plugin- a state streaming plugin.
+We will define a `StateStreaming` plugin interface which extends the above `Plugin` interface to support a state streaming service.
 
-// ServiceTypeFromString returns the streaming.ServiceConstructor corresponding to the provided name
-func ServiceTypeFromString(name string) (ServiceConstructor, error) {
-  ssType := NewStreamingServiceType(name)
-  if ssType == Unknown {
-    return nil, fmt.Errorf("unrecognized streaming service name %s", name)
-  }
-  if constructor, ok := ServiceConstructorLookupTable[ssType]; ok {
-    return constructor, nil
-  }
-  return nil, fmt.Errorf("streaming service constructor of type %s not found", ssType.String())
-}
+```go
+// StateStreaming interface for plugins that load a baseapp.StreamingService onto a baseapp.BaseApp
+type StateStreaming interface {
+	// Register configures and registers the plugin streaming service with the BaseApp
+	Register(bApp *baseapp.BaseApp, marshaller codec.BinaryCodec, keys map[string]*types.KVStoreKey) error
 
-// NewFileStreamingService is the streaming.ServiceConstructor function for creating a FileStreamingService
-func NewFileStreamingService(opts serverTypes.AppOptions, keys []sdk.StoreKey, marshaller codec.BinaryMarshaler) (sdk.StreamingService, error) {
-  filePrefix := cast.ToString(opts.Get("streamers.file.prefix"))
-  fileDir := cast.ToString(opts.Get("streamers.file.write_dir"))
-  return file.NewStreamingService(fileDir, filePrefix, keys, marshaller)
+	// Start starts the background streaming process of the plugin streaming service
+	Start(wg *sync.WaitGroup)
+
+	// Plugin is the base Plugin interface
+	Plugin
 }
```

Comment:
shouldn't the `StateStreaming` have a writer methods?

Reply:
`StateStreaming` plugin uses the `Register` method to register `StreamingService`(s) with the `BaseApp`, the `StreamingService` interface exposes `WriteListener`(s) that have the `OnWrite` method. So in other words, the writing interface is buried two levels down. Let me see if I can restructure this so that it is more clear where writing is occurring.